### DISCUSSION TITLE
High CPU usage on CRDB and very slow environment

### DISCUSSION THREAD

**Discussion by danielmichalichyn** [zitadel/zitadel#7138] [Q&A]
Hello guys, 

I need a little help to identify a high CPU usage in CockroachDB. I'm currently running version 2.41.1, in a single instance of CockroachDB and Zitadel on docker swarm.

We don't have a lot of users right now, the only significant application using Zitadel is an API that constantly requests for token impersonation of service users PAT, so i can check the roles associated with that service user and to see if the PAT is still active. 

When our API receive a lot of requests in a small time window, Zitadel and Cockroach starts to become really slow and consume a lot of CPU. But honestly i don't know the approach to debug this. Can anyone help me?

![image](https://github.com/zitadel/zitadel/assets/52325588/c1a3ba88-8b70-47b0-ac23-8363f6c30c6f)
![image](https://github.com/zitadel/zitadel/assets/52325588/575bfcc6-d7ba-4e11-8e77-043d68cc75d6)

**Comment by hifabienne** [comment]
@adlerhurst @muhlemmer can you help here?

**Comment by adlerhurst** [comment]
Huhu

We made a lot of performance improvements in releases >= v2.42.4. Can you try one of those?

**Reply by danielmichalichyn** [reply]
Ok, i'll try to update to this version and check if the app performance will be better.  I've activated the debug level on my Zitadel instance, and i'm just wandering if it's normal to see the projections continuously running?

This trigger iteration for projections never stop. And there is times that projections fail with the following message:
"unable to filter events Parent=(context deadline exceeded)"

![image](https://github.com/zitadel/zitadel/assets/52325588/cc4af954-e96b-40f2-a371-add0035e1d7e)

**Reply by adlerhurst** [reply]
Yes they are scheduled. The goal is to process events which might not have been processed yet. That error can happen from time to time, especially when there are multiple containers of zitadel running or events are written at the same time the scheduler is executed

**Reply by danielmichalichyn** [reply]
Hello @adlerhurst , how are you?
I've updated the Zitadel version to the latest release 2.43.5. As far as i'm understanding correctly, the CPU usage should not be a problem, this seems to be the Cockroach doing it's own thing. But i'm noticing a strange behaviour by the Zitadel application that looks like a memory leak. I'm hosting Zitadel on a docker swarm, but my Zitadel container is running on a host dedicated for him and from time to time the container restarts by a OOM kill. Following is the memory usage graph for that specific host. Can you help me debug this and identify if there is a potencial memory leak?

![image](https://github.com/zitadel/zitadel/assets/52325588/f4b5117b-88b1-4c05-ade3-201238ce702d)

**Reply by muhlemmer** [reply]
Interesting problem. Can you please open a bug report for it and mention me and @adlerhurst.

I will try to have a look after the weekend.
